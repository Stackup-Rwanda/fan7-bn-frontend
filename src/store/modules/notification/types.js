export const GET_NOTIFICATIONS_START = 'GET_NOTIFICATIONS_START';
export const GET_NOTIFICATIONS_SUCCESS = 'GET_NOTIFICATIONS_SUCCESS';
export const GET_NOTIFICATIONS_ERROR = 'GET_NOTIFICATIONS_ERROR';

export const MARK_NOTIFICATION_AS_READ_START = 'MARK_NOTIFICATION_AS_READ_START';
export const MARK_NOTIFICATION_AS_READ_SUCCESS = 'MARK_NOTIFICATION_AS_READ_SUCCESS';
export const MARK_NOTIFICATION_AS_READ_ERROR = 'MARK_NOTIFICATION_AS_READ_ERROR';

export const MARK_ALL_NOTIFICATIONS_AS_READ_START = 'MARK_ALL_NOTIFICATIONS_AS_READ_START';
export const MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS = 'MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS';
export const MARK_ALL_NOTIFICATIONS_AS_READ_ERROR = 'MARK_ALL_NOTIFICATIONS_AS_READ_ERROR';
