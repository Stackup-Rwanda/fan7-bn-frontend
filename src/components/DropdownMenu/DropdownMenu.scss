.dropdown-menu {
  position: relative;

  &__button {
    display: block;
    font-weight: normal;
    text-align: center;
    white-space: nowrap;
    background-color: transparent;
    background-image: none;
    -ms-touch-action: manipulation;
    touch-action: manipulation;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: none;
    cursor: pointer;

    &:focus,
    &:active &:active:focus {
      outline: 0;
      background-image: none;
    }
  }

  &__icon {
    height: 40px;
    width: 40px;
    cursor: pointer;
    border: none;
  }

  &__text {
    position: absolute;
    background: red;
    color: #ffffff;
    width: 25px;
    height: 25px;
    padding: 0.25em;
    border-radius: 50%;
    margin-left: -20px;
    text-align: center;
    vertical-align: middle;
    /**/
    cursor: pointer;
    border: none;
    font-weight: bold;
  }

  &__content {
    display: none;
    position: absolute;
    left: -200px;
    z-index: 9999;
    background: #f9f9f9;
    min-width: 80px;
    max-height: 600px;
    overflow: auto;
    box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
  }
}

.show {
  display: block;
}
