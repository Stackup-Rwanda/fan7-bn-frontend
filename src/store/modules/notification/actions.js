import {
  GET_NOTIFICATIONS_START,
  GET_NOTIFICATIONS_SUCCESS,
  GET_NOTIFICATIONS_ERROR,
  MARK_NOTIFICATION_AS_READ_START,
  MARK_NOTIFICATION_AS_READ_SUCCESS,
  MARK_NOTIFICATION_AS_READ_ERROR,
  MARK_ALL_NOTIFICATIONS_AS_READ_START,
  MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS,
  MARK_ALL_NOTIFICATIONS_AS_READ_ERROR,
} from './types';
import HttpService from '../../../utils/HttpService';
import Errors from '../../../utils/helpers/errors';

export const getNotifications = () => async (dispatch) => {
  dispatch({ type: GET_NOTIFICATIONS_START });
  try {
    const res = await HttpService.get('/notifications');
    dispatch({ type: GET_NOTIFICATIONS_SUCCESS, payload: res.data });
  } catch (error) {
    dispatch({ type: GET_NOTIFICATIONS_ERROR, payload: Errors.selectMessage(error) });
  }
};

export const markOneAsRead = (notificationId) => async (dispatch) => {
  dispatch({ type: MARK_NOTIFICATION_AS_READ_START });
  try {
    const res = await HttpService.patch(`/notifications/${notificationId}/read`);
    dispatch({ type: MARK_NOTIFICATION_AS_READ_SUCCESS });
  } catch (error) {
    Errors.handle(error);
    dispatch({ type: MARK_NOTIFICATION_AS_READ_ERROR, payload: Errors.selectMessage(error) });
  }
};

export const markAllAsRead = () => async (dispatch) => {
  dispatch({ type: MARK_ALL_NOTIFICATIONS_AS_READ_START });
  try {
    const res = await HttpService.patch('/notifications');
    dispatch({ type: MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS });
  } catch (error) {
    Errors.handle(error);
    dispatch({ type: MARK_ALL_NOTIFICATIONS_AS_READ_ERROR, payload: Errors.selectMessage(error) });
  }
};
