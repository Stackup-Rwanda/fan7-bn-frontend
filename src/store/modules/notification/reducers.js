import {
  GET_NOTIFICATIONS_START,
  GET_NOTIFICATIONS_SUCCESS,
  GET_NOTIFICATIONS_ERROR,
  MARK_NOTIFICATION_AS_READ_START,
  MARK_NOTIFICATION_AS_READ_SUCCESS,
  MARK_NOTIFICATION_AS_READ_ERROR,
  MARK_ALL_NOTIFICATIONS_AS_READ_START,
  MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS,
  MARK_ALL_NOTIFICATIONS_AS_READ_ERROR,
} from './types';

const initialState = {
  loading: false,
  notifications: [],
  error: null,
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case GET_NOTIFICATIONS_START:
      return { ...state, loading: true };

    case GET_NOTIFICATIONS_SUCCESS:
      return { ...state, loading: false, notifications: payload };

    case GET_NOTIFICATIONS_ERROR:
      return { ...state, loading: false, error: payload };

    case MARK_NOTIFICATION_AS_READ_START:
      return { ...state, loading: true };

    case MARK_NOTIFICATION_AS_READ_SUCCESS:
      return { ...state, loading: false };

    case MARK_NOTIFICATION_AS_READ_ERROR:
      return { ...state, loading: false, error: payload };

    case MARK_ALL_NOTIFICATIONS_AS_READ_START:
      return { ...state, loading: true };

    case MARK_ALL_NOTIFICATIONS_AS_READ_SUCCESS:
      return { ...state, loading: false };

    case MARK_ALL_NOTIFICATIONS_AS_READ_ERROR:
      return { ...state, loading: false, error: payload };

    default:
      return state;
  }
};
