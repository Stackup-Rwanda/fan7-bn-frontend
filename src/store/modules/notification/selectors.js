import { createSelector } from 'reselect';

export const selectRaw = (notifications) => notifications;

export const selectNotifications = createSelector([selectRaw], (notifications) =>
  notifications
    ? notifications.notifications.filter((notification) => notification.status === 'unread')
    : []
);

export const selectCount = createSelector([selectNotifications], (notifications) =>
  notifications ? notifications.length : 0
);

export const selectLoading = createSelector([selectRaw], (notifications) =>
  notifications ? notifications.loading : false
);
