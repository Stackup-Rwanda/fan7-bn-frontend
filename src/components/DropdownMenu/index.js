import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './DropdownMenu.scss';
import notificationIcon from '../../assets/icons/notification.png';

class DropdownMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };
    this.toggleMenu = this.toggleMenu.bind(this);
  }

  // componentWillMount() {
  //   window.addEventListener('click', function (e) {
  //     this.setState({ isOpen: false });
  //   });
  // }

  toggleMenu(e) {
    e.preventDefault();
    this.setState({ isOpen: !this.state.isOpen });
  }

  render() {
    const { children, icon, text } = this.props;

    if (children.length === 0) {
      throw 'DropdownMenu must have at least one MenuItem child.';
    }

    return (
      <div className="dropdown-menu">
        <button type="button" className="dropdown-menu__button" onClick={this.toggleMenu}>
          {icon && <img src={icon} className="dropdown-menu__icon" />}
          {text ? (
            <span className="dropdown-menu__text">{text}</span>
          ) : (
            <span className="dropdown-menu__text">0</span>
          )}
        </button>
        <div
          className={this.state.isOpen ? 'dropdown-menu__content show' : 'dropdown-menu__content'}
        >
          {children}
        </div>
      </div>
    );
  }
}

DropdownMenu.propTypes = {
  icon: PropTypes.string,
  text: PropTypes.string,
  children: PropTypes.node.isRequired,
};

DropdownMenu.defaultProps = {
  icon: notificationIcon,
  text: '',
};

export default DropdownMenu;
