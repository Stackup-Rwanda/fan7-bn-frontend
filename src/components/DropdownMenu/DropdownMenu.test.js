/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import DropdownMenu from './';
import notification from '../../assets/icons/notification.png';

describe('DropdownMenu Component Test', () => {
//   it('should throw an error if no children', () => {
//     const props = {
//       children: [],
//     };

//     const wrapper = shallow(<DropdownMenu {...props} />);

//     expect(() => {
//       wrapper;
//     }).toThrow('DropdownMenu must have at least one MenuItem child.');
//   });

  it('should return true if component exists', () => {
    const props = {
      icon: notification,
    };

    const wrapper = shallow(
      <DropdownMenu {...props}>
        <p>Menu item</p>
      </DropdownMenu>
    );

    expect(wrapper.find('.dropdown-menu').exists()).toBe(true);
  });

  it('should render icon', () => {
    const props = {
      icon: notification,
    };

    const wrapper = shallow(
      <DropdownMenu {...props}>
        <p>Menu item</p>
      </DropdownMenu>
    );

    expect(wrapper.find('.dropdown-menu__icon').exists()).toBe(true);
  });

  it('should render text', () => {
    const props = {
      text: 'Menu',
    };

    const wrapper = shallow(
      <DropdownMenu {...props}>
        <p>Menu item</p>
      </DropdownMenu>
    );

    expect(wrapper.find('.dropdown-menu__text').exists()).toBe(true);
  });

  it('should toggle menu', () => {
    const props = {
      icon: notification,
    };
    const event = {
      preventDefault: jest.fn(),
    };

    const wrapper = shallow(
      <DropdownMenu {...props}>
        <p>Menu item</p>
      </DropdownMenu>
    );

    wrapper.find('.dropdown-menu__button').simulate('click', event);
    expect(wrapper.find('.show').exists()).toBe(true);
  });
});

describe('<DropdownMenu /> shallow rendering tests', () => {
  it('matches the snapshot', () => {
    const props = {
      icon: notification,
    };
    const tree = shallow(
      <DropdownMenu {...props}>
        <p>Menu item</p>
      </DropdownMenu>
    );
    expect(tree).toMatchSnapshot();
  });
});
