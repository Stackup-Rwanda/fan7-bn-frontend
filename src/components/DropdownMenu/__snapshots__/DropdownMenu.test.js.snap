// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownMenu /> shallow rendering tests matches the snapshot 1`] = `
<div
  className="dropdown-menu"
>
  <button
    className="dropdown-menu__button"
    onClick={[Function]}
    type="button"
  >
    <img
      className="dropdown-menu__icon"
      src={
        Object {
          "process": [Function],
        }
      }
    />
    <span
      className="dropdown-menu__text"
    >
      0
    </span>
  </button>
  <div
    className="dropdown-menu__content"
  >
    <p>
      Menu item
    </p>
  </div>
</div>
`;
